crates/pallet-grandpa-finality-verifier/src/lib.rs

```diff
@@ -146,32 +145,18 @@ pub mod pallet {
         let block = C::decode_block::<T>(object)?;
         let number = *block.block.header.number();
         let hash = block.block.header.hash();
+        let current_height = CurrentBlockHeight::<T>::get(chain_id);
 
-        match BestFinalized::<T>::get(chain_id) {
-            None => {
-                // this is a first block we are importing
-                // ensure it is before the checkpoint
-                ensure!(
-                    (*block.block.header.number()).into()
-                        <= ValidationCheckpoint::<T>::get(chain_id),
-                    Error::<T>::InvalidHeader
-                );
-                BestFinalized::<T>::insert(chain_id, block.block.header.encode());
-                return Ok((hash, number));
-            }
-            Some(data) => {
-                let best_finalized = C::decode_header::<T>(data.as_slice())?;
-                // ensure block is always increasing
-                let next_block_number = best_finalized
-                    .number()
-                    .checked_add(&One::one())
-                    .ok_or(sp_runtime::ArithmeticError::Overflow)?;
-                ensure!(next_block_number == number, Error::<T>::InvalidHeader);
-            }
-        };
+        // ensure block is always incrementing by 1
+        ensure!(
+            number.into() == current_height + 1,
+            Error::<T>::InvalidHeader
+        );
 
-        // check if we should validate or just import
-        if (*block.block.header.number()).into() > ValidationCheckpoint::<T>::get(chain_id) {
+        // if the target header is a descendent of our best known finalized header, validate the justification and import
+        let best_finalized =
+            C::decode_header::<T>(BestKnownFinalized::<T>::get(chain_id).as_slice())?;
+        if number > *best_finalized.number() {
```

Comment:
I think 2 things are left here:
1) Block body check, I only see header is read, but body might be malformed too
2) In `number == *best_finalized.number()` case we need to ensure header of the block is exactly the same as we have stored as best finalized header

Reply:
> Block body check, I only see header is read, but body might be malformed too

This is done here - https://github.com/subspace/subspace/pull/380/files#diff-321bf7696e02a969592d8a5687871d5f6fb44fa5a0ba104d1b4aab8ba13abe03R145

> In number == *best_finalized.number() case we need to ensure header of the block is exactly the same as we have stored as best finalized header

Like a sanity check ?
Hmmm okay

Reply:
> Block body check, I only see header is read, but body might be malformed too

Or were you expecting that we also store the latest block itself on the state ?

Reply:
> This is done here

It does decoding, but it doesn't verify that the contents of the transactions correspond to the extrinsics root in the header, right?

> Like a sanity check ?

We can't just allow to submit anything for that specific block, we need to verify it properly.

> Or were you expecting that we also store the latest block itself on the state ?

No